Feature/azure whisper

Sheepsta300 opened the pull request:
Here is my first attempt at creating an AzureOpenAIWhisperParser:

- Modelled after the OpenAIWhisperParser.

- No tests yet.

Sheepsta300 commented:
Not quite sure how to make check 3.8 work as it says it ran out of disk space?

Sheepsta300 commented:
I've tested to see how the class works when no env variables are set and it works as expected. It checks using os.environ and will throw appropriate errors for each value if it can't find them.

I added a few comments in the class to clarify a bit what needs to be set/what env variables are expected.

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+    This is different to the Open AI Whisper parser and requires
+    an Azure OpenAI API Key.
+
+    Args:
```
I'm not sure if this is aligned to the docstring formatting used elsewhere - though I don't think this formatting is sphinx-compatible... but then maybe they're not using sphinx. So, if this is the formatting used for other integrations this that's all good.

Sheepsta300 replied:
I've written up a better docstring I think covers most of what the class does and I've tried to format it similar to other azure services, let me know though if you think anything needs to be added/removed :)

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+
+    Args:
+        api_key: Azure OpenAI API key
+        chunk_duration_threshold: minimum duration of a chunk in seconds
```
```suggestion
        chunk_duration_threshold: Minimum duration of a chunk in seconds.
```

Sheepsta300 replied:
Updated

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+    Args:
+        api_key: Azure OpenAI API key
+        chunk_duration_threshold: minimum duration of a chunk in seconds
+            NOTE: According to the OpenAI API, the chunk duration should be at least 0.1
```
Does this guidance hold true for Azure as well?

Sheepsta300 replied:
```python
Transcribing part 1!
Attempt 1 failed. Exception: Error code: 400 - {'error': {'message': 'Audio file is too short. Minimum audio length is 0.1 seconds.', 'type': 'invalid_request_error', 'param': 'file', 'code': 'audio_too_short'}}
Attempt 2 failed. Exception: Error code: 400 - {'error': {'message': 'Audio file is too short. Minimum audio length is 0.1 seconds.', 'type': 'invalid_request_error', 'param': 'file', 'code': 'audio_too_short'}}
```

After using audio less than the minimum chunk size it does indeed throw an error, so this still holds true for the Azure Whisper API.

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+        ] = None,
+        temperature: Union[float, None] = None,
+    ):
+        self.api_key = api_key
```
We don't support env var here?

Sheepsta300 replied:
env var is now supported here the same as for azure_endpoint and version

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+    ):
+        self.api_key = api_key
+        self.base_url = base_url or os.environ.get("AZURE_OPENAI_ENDPOINT")
+        self.api_version = api_version or os.environ.get("OPENAI_API_VERSION")
```
Is it the standard convention for parsers that all these attributes be public and settable?

Sheepsta300 replied:
As far as I can tell yes. Looking over other files and parsers this seems to be a common trend, maybe to allow users easier access.

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+
+    def __init__(
+        self,
+        api_key: Optional[str] = None,
```
Any reason this shouldn't be keyword-only as well?

Sheepsta300 replied:
Woops, my bad! wrong placement

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+    """Transcribe and parse audio files.
+
+    Audio transcription is with Azure OpenAI Whisper model.
+    This is different to the Open AI Whisper parser and requires
```
Do we need to mention here that although it's different than OpenAI, we're still using the openai SDK (and is therefore still a dependency)? It might also be worth elaborating how it's different to Open AI - for example, behavioural differences, payload formatting etc

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+        self.temperature = temperature
+
+    @property
+    def _create_params(self) -> Dict[str, Any]:
```
Is there an existing convention that this should be an `@property` rather than a method?

Sheepsta300 replied:
I haven't found anything on LangChain that specifies if properties should be used over methods or vice versa, but what would your opinion be?

I feel like it makes sense to have a property here as it's essentially a `get` method with a filter for any `None` values, meaning theres no complex computation going on in the property

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+            # Skip chunks that are too short to transcribe
+            if chunk.duration_seconds <= self.chunk_duration_threshold:
+                continue
+            file_obj = io.BytesIO(chunk.export(format="mp3").read())
```
Should we be hard coding this to mp3? Is format something that we should allow users to configure?
What will the behaviour be if we set format=mp3, when the data is encoded with something else?

Sheepsta300 replied:
I've changed this section to us `os.path.splitext` to get the file extension from the file and use it to format the file_obj instead of hard code it.

I also added a file checking section to the code:
```python
if not str(blob.path).endswith(self.input_format):
            raise ValueError(
                "File must be of one of the following types: "
                f"{str(self.input_format)}"
            )
```

`input_format` is a Tuple containing the allowed file types set by the Whisper API. I did this because the error returned when the file is the wrong type wasn't very informative and I thought this might work better. I also thought it might be good because the user could configure the `input_format` to only allow certain file types, if for whatever reason they wanted to only use say MP3 and MP4.

Let me know if this makes sense or if it should be removed.

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+        audio = AudioSegment.from_file(blob.path)
+
+        # Define the duration of each chunk in minutes
+        # Need to meet 25MB size limit for Whisper API
```
It could be worth making these kinds of limits configurable - as such size limits do have a tendency to change on the service over time.

Sheepsta300 replied:
I'm not really sure how I feel about this part of the code. The limit that was set in the OpenAIWhisperParser allows for 20 minutes of content to be parsed:

```python
        # Need to meet 25MB size limit for Whisper API
        chunk_duration = 20
        chunk_duration_ms = chunk_duration * 60 * 1000

        # Split the audio into chunk_duration_ms chunks
        for split_number, i in enumerate(range(0, len(audio), chunk_duration_ms)):
```

This doesn't actually enforce the 25 MB limit and 20 minutes of content is also quite a lot, most likely bigger than 25 MB anyway.

Since the Azure Whisper parser has a maximum of 25MB and will throw an error, it may be better just to remove the file chunking/splitting part of the code. The Azure AI Speech has a [batch transcribe](https://learn.microsoft.com/azure/ai-services/speech-service/batch-transcription-create?pivots=rest-api#use-a-whisper-model) which the documentation recommends to use for larger files. Because of this I'm not sure it really fits the purpose of the Azure Whisper model to be splitting larger files in order to transcribe them?

Let me know what you think anyway!

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+                continue
+            file_obj = io.BytesIO(chunk.export(format="mp3").read())
+            if blob.source is not None:
+                file_obj.name = blob.source + f"_part_{split_number}.mp3"
```
Does this mean the `blob.source` will never already have a file-extension?

Sheepsta300 replied:
I've changed this to us `os.path.splitext` and use the first element. This will ensure even if it has an extension or not, it will not be included.

annatisch commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+                            self.deployment_id, file_obj
+                        )
+                    break
+                except Exception as e:
```
We should be more specific here on which exceptions we retry on. Retrying on an Authentication error for example is unlikely to help. Or if one of the inputs is invalid and raises something like a TypeError. The only cases we want to retry are cases of service-side throttling or intermittent connectivity issues - everything else should just get raised.

Sheepsta300 replied:
I've removed the while loop retries/attempts and added a `max_retries` attribute set to a default of 3 (this can also be set by the user). This should handle retrying for any connectivity or throttling issues without throwing an error, all other Exceptions will just get raised.

annatisch reviewed (commented):
Thanks for this! It seems that a lot of the things I've commented on are largely from the original OpenAI parser implementation - but we can still look into whether there's some improvements to be made and then maybe even contribute back to the OpenAI implementation. ;)

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+            it will be skipped.
+    """
+
+    def __init__(
```
Maybe not for the initial PR, but given this is the Azure version for whisper, I think would be great to also support Entra ID token authentication along with API key. This will probably look like supporting an `azure_ad_token_provider` parameter similar to how it's done in AzureOpenAI: https://github.com/langchain-ai/langchain/blob/44e3e2391c48bfd0a8e6a20adde0b6567f4f43c3/libs/partners/openai/langchain_openai/llms/azure.py#L59

Sheepsta300 replied:
I've added an attribute `azure_ad_token` which is set to `None` as default. There's no problem I've found if it gets passed to  `AzureOpenAI` as it's default (None), so it can just be set when need be.

kristapratico replied:
Can I suggest we support `azure_ad_token_provider` instead? While the AzureOpenAI client does accept `azure_ad_token`, it's less useful since it's authenticating with a static token that is set to expire in ~1 hour. If someone uses this in their app, they will start seeing failures once the token is invalid. With the token provider, the user can pass a callable that will refresh the token value before it expires. 

You can type it like this in the whisper parser:

 azure_ad_token_provider: Union[Callable[[], str], None] = None

and the common way that users will use this is to pass in the helper function from our azure.identity library, like:

```python
from azure.identity import DefaultAzureCredential, get_bearer_token_provider

from openai import AzureOpenAI

token_provider = get_bearer_token_provider(DefaultAzureCredential(), "https://cognitiveservices.azure.com/.default")

AzureOpenAI(..., azure_ad_token_provider=token_provider)
```

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+        self,
+        api_key: Optional[str] = None,
+        *,
+        deployment_id: str,
```
We should try to stay consistent with the parameter naming for an Azure deployment across all the different langchain classes. Maybe we should use `deployment_name` or `azure_deployment`?

Sheepsta300 replied:
Now the attribute is called `deployment_name` :)

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,6 +12,144 @@
+        *,
+        deployment_id: str,
+        chunk_duration_threshold: float = 0.1,
+        base_url: Optional[str] = None,
```
Perhaps we rename `base_url` to `azure_endpoint` to better match what Azure is expecting?

Sheepsta300 replied:
I have updated this to use `azure_endpoint` now`.

Sheepsta300 commented:
Just removed the chunking from the class and I added a section to check if the file is below 25 MB. I tried inputting a large file (over 25MB) and it took about 7 minutes for the API to return and throw an error so I figured it made more sense to check before starting the transcription. 

Also not sure again why the CI has not worked! It says the `numpy` version is unsatisfiable, but I haven't included/ don't use `numpy` in this class.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+    """
+    Transcribe and parse audio files using Azure OpenAI Whisper.
+
+    This parser integrates with the Azure OpenAI Whisper model to transcribe
+    audio files. It differs from the standard OpenAI Whisper parser, requiring
+    an Azure API key and endpoint. The parser is limited to files under 25 MB.
```
Since we plan to support more than just API key, I think we can say Azure endpoint and credentials.

Sheepsta300 replied:
updated.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+    an Azure API key and endpoint. The parser is limited to files under 25 MB.
+
+    **Note**:
+    While this parser uses the OpenAI API, it is configured for Azure OpenAI
+    services. The key difference is the integration with the Azure ecosystem,
+    making it suitable for workflows involving other Azure services.
```
The parser uses the Azure OpenAI API

Sheepsta300 replied:
updated.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+
+        .. code-block:: bash
+
+            pip install -U langchain
```
since this is in community, do I also need to install langchain-community?

Sheepsta300 replied:
my bad, I thought it could be accessed just by installing LangChain, but after checking, that's not true, has been updated to:

```python
pip install -U langchain langchain-community
```

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+        response_format: Union[
+            Literal["json", "text", "srt", "verbose_json", "vtt"], None
+        ] = None,
+        input_format: Tuple = (
+            ".flac",
+            ".mp3",
+            ".mp4",
+            ".mpeg",
+            ".mpga",
+            ".m4a",
+            ".ogg",
+            ".wav",
+            ".webm",
+        ),
```
I would recommend not adding `input_format`. The Azure OpenAI service may support new formats and if they do, we'll need to update this code each time to include it. We generally favor doing less client side validation and letting the service validate the request / raise an error if an unsupported format is passed.

Sheepsta300 replied:
Removed.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+                "pydub package not found, please install it with " "`pip install pydub`"
+            )
+
+        if not str(blob.path).endswith(self.input_format):
+            raise ValueError(
+                "File must be of one of the following types: "
+                f"{str(self.input_format)}"
+            )
+
+        if os.path.getsize(str(blob.source)) > self.byte_limit:
+            raise ValueError(
+                "Audio file exceeds 25 MB limit,"
+                f" length of file is {os.path.getsize(str(blob.source))} bytes"
+            )
```
Same comment here about client side validation, let's allow the service decide if the request doesn't meet supported parameters.

Sheepsta300 replied:
```python
        if os.path.getsize(str(blob.source)) > self.byte_limit:
            raise ValueError(
                "Audio file exceeds 25 MB limit,"
                f" length of file is {os.path.getsize(str(blob.source))} bytes"
            )
```
I included this chunk as when a large file is passed to the api it takes quite a while for an error to be thrown. I have an idea that the transcription starts and the error will be thrown when it hits a limit as opposed to checking beforehand. Let me know if still don't think it's worth it or if there are any better ways to implement this.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+                azure_ad_token=self.azure_ad_token,
+            )
+        else:
+            # Set the API key if provided
```
if not using openai>=1, we also need to set api_type and api_version here:

openai.api_type = "azure"
openai.api_base = self.azure_endpoint
openai.api_key = self.api_key
openai.api_version = self.api_version

also, I don't believe openai<1 had a `base_url` property. This might be a bug in the openai whisper parser

Sheepsta300 replied:
changed to 
```
        else:
            if self.api_key:
                openai.api_key = self.api_key
            if self.azure_endpoint:
                openai.api_base = self.azure_endpoint
            if self.api_version:
                openai.api_version = self.api_version
            openai.api_type = "azure"
```
I didn't realise this was wrong, would it make sense to suggest a change to the openai version then also?

kristapratico replied:
We can update the openai.base_url to openai.api_base for the openai parser.

Sheepsta300 replied:
Sure thing! I'll make a note of that when I make the final PR too.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+                    **self._create_params,
+                )
+            else:
+                transcript = openai.Audio.transcribe(self.deployment_name, file_obj)
```
The <v1 library also requires deployment_id to be passed for Azure (it's the same value as model)

```suggestion
                transcript = openai.Audio.transcribe(model=self.deployment_name, deployment_id=self.deployment_name, file=file_obj)
```

Sheepsta300 replied:
updated.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+        # Get file extension from the file path
+        dot_file_extension = os.path.splitext(str(blob.path))[1]
+        file_extension = dot_file_extension[1:]
+        # m4a and mpga are not valid format flags for ffmpeg
+        # They do have equivalent flags that can be used though
+        format_map = {"m4a": "mp4", "mpga": "mp3"}
```
does the service raise an exception for these formats?

Sheepsta300 replied:
This actually had to do with pythons `io` library, it doesn't support these formats

```python
        format_map = {"m4a": "mp4", "mpga": "mp3"}
        file_extension = format_map.get(file_extension, file_extension)

        file_obj = io.BytesIO(audio.export(format=file_extension).read())
```
But I realised since we're not chunking larger files anymore and letting the API throw an error if the file is too large, this whole chunk can be taken out, it's now:

```python
        file_obj = open(blob.path, 'rb')

        #Transcribe
        try:
            if is_openai_v1():
                transcript = client.audio.transcriptions.create(
                    model=self.deployment_name,
                    file=file_obj,
                    **self._create_params,
                )
            else:
                transcript = openai.Audio.transcribe(self.deployment_name, file_obj)
        except Exception:
            raise
```

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,273 @@
+            raise
+
+        yield Document(
+            page_content=transcript.text,
```
I think we need this same check here as in the openai whisper parser:

page_content=transcript.text
                if not isinstance(transcript, str)
                else transcript,


It seems weird, but if a user selects response_format="text", the transcript returned will be `str` instead of an object with a text property

Sheepsta300 replied:
Did not know that! has been updated to:

```python
yield Document(
                page_content=transcript.text
                if not isinstance(transcript, str)
                else transcript,
                metadata={"source": blob.source},
            )
```

kristapratico reviewed (commented):
This is looking really good! I think the next step is to add some tests :)

I suggest taking a look at the contributing guide to understand how to set up / write tests for the langchain repo: https://python.langchain.com/v0.2/docs/contributing/

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+    **Note**:
+    While this parser uses the Azure OpenAI API, it is configured for Azure OpenAI
+    services. The key difference is the integration with the Azure ecosystem,
+    making it suitable for workflows involving other Azure services.
```
I think we should re-word this a little:

```suggestion
    This parser uses the Azure OpenAI API, providing integration with the Azure ecosystem, and
    making it suitable for workflows involving other Azure services.
```

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+
+            whisper_parser = AzureOpenAIWhisperParser(
+                deployment_name="your-whisper-deployment",
+                api_version="2024-05-01-preview",
```
Let's select a non-preview API version for the example

```suggestion
                api_version="2024-06-01",
```

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+        api_key: Optional[str] = None,
+        azure_endpoint: Optional[str] = None,
+        api_version: Optional[str] = None,
+        azure_ad_token: Optional[str] = None,
```
Can we add this parameter?

azure_ad_token_provider: Union[Callable[[], str], None] = None,

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+        self.azure_endpoint = azure_endpoint or os.environ.get("AZURE_OPENAI_ENDPOINT")
+        self.api_version = api_version or os.environ.get("OPENAI_API_VERSION")
+        self.azure_ad_token = azure_ad_token
+
```
self.azure_ad_token_provider = azure_ad_token_provider

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+                azure_endpoint=self.azure_endpoint,
+                api_version=self.api_version,
+                max_retries=self.max_retries,
+                azure_ad_token=self.azure_ad_token,
```
azure_ad_token_provider=self.azure_ad_token_provider

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+        ] = None,
+        temperature: Optional[float] = None,
+        deployment_name: str,
+        chunk_duration_threshold: float = 0.1,
```
Given we no longer use this parameter I think we can safely remove it.

kristapratico commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -12,14 +13,232 @@
+                transcript = self._client.Audio.transcribe(
+                    model=self.deployment_name,
+                    deployment_id=self.deployment_name,
+                    file=file_obj,
```
I think we need to pass the **self._create_params here as well

kristapratico reviewed (commented):
Few more small things, after which I think we're ready to officially PR this one as well. :)

Sheepsta300 commented:
Finally got 3 unit tests working for the parser! woooo

Sheepsta300 commented:
Will now quickly work on making a doc for the Parser to fit contribution guidelines
